datafusion-federation/src/optimize/optimize_projections/mod.rs

```diff
@@ -0,0 +1,1047 @@
+/// * `columns` - A mutable reference to a `HashSet<Column>` where detected
+///   columns are collected.
+fn outer_columns_helper_multi<'a, 'b>(
+    exprs: impl IntoIterator<Item = &'a Expr>,
+    columns: &'b mut HashSet<&'a Column>,
+) {
+    exprs.into_iter().for_each(|e| outer_columns(e, columns));
+}
+
+/// Splits requirement indices for a join into left and right children based on
+/// the join type.
+///
+/// This function takes the length of the left child, a slice of requirement
+/// indices, and the type of join (e.g. `INNER`, `LEFT`, `RIGHT`) as arguments.
+/// Depending on the join type, it divides the requirement indices into those
+/// that apply to the left child and those that apply to the right child.
+///
+/// - For `INNER`, `LEFT`, `RIGHT` and `FULL` joins, the requirements are split
+///   between left and right children. The right child indices are adjusted to
+///   point to valid positions within the right child by subtracting the length
+///   of the left child.
+///
+/// - For `LEFT ANTI`, `LEFT SEMI`, `RIGHT SEMI` and `RIGHT ANTI` joins, all
+///   requirements are re-routed to either the left child or the right child
+///   directly, depending on the join type.
+///
+/// # Parameters
+///
+/// * `left_len` - The length of the left child.
+/// * `indices` - A slice of requirement indices.
+/// * `join_type` - The type of join (e.g. `INNER`, `LEFT`, `RIGHT`).
+///
+/// # Returns
+///
+/// A tuple containing two vectors of `usize` indices: The first vector represents
+/// the requirements for the left child, and the second vector represents the
+/// requirements for the right child. The indices are appropriately split and
+/// adjusted based on the join type.
+fn split_join_requirements(
+    left_len: usize,
+    indices: RequiredIndicies,
+    join_type: &JoinType,
+) -> (RequiredIndicies, RequiredIndicies) {
+    match join_type {
+        // In these cases requirements are split between left/right children:
+        JoinType::Inner
+        | JoinType::Left
+        | JoinType::Right
+        | JoinType::Full
+        | JoinType::LeftMark => {
+            // Decrease right side indices by `left_len` so that they point to valid
+            // positions within the right child:
+            indices.split_off(left_len)
+        }
+        // All requirements can be re-routed to left child directly.
+        JoinType::LeftAnti | JoinType::LeftSemi => (indices, RequiredIndicies::new()),
+        // All requirements can be re-routed to right side directly.
+        // No need to change index, join schema is right child schema.
+        JoinType::RightSemi | JoinType::RightAnti => (RequiredIndicies::new(), indices),
+    }
+}
+
+fn add_projection_on_top_if_helpful(
+    plan: LogicalPlan,
+    _project_exprs: Vec<Expr>,
+    _config: &dyn OptimizerConfig,
+) -> Result<Transformed<LogicalPlan>> {
+    // Always prefer the original query layout.
+    Ok(Transformed::no(plan))
```

Comment:
Alternatively we can remove usage of `add_projection_on_top_if_helpful` completely but this approaches simplify future updates as we modify few helper functions only w/o changing core logic => easy to upgrade moving forward (if it will be erquired)